lib/en/man1/aerb.1
'\"tp
.\" aegis - project change supervisor
.\" Copyright (C) 2001, 2002, 2005-2008, 2010, 2012 Peter Miller
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program. If not, see <http://www.gnu.org/licenses/>.
.\"
.so lib/en/man1/z_name.so
.TH "\*(n) \-Review_Begin" 1 \*(N) "Reference Manual"
.SH NAME
aegis review begin \- begin a change review
.XX "aerb(1)" "begin a change review"
.SH SYNOPSIS
.B \*(n)
.B \-Review_Begin
[
.IR option ...
]
.br
.B \*(n)
.B \-Review_Begin
.B \-List
[
.IR option ...
]
.br
.B \*(n)
.B \-Review_Begin
.B \-Help
.SH DESCRIPTION
The
.I \*(n)
.I \-Review_Begin
command is used to
notify \*(n) that you have begun to review a change.
.PP
The change will be advanced from the
.I "awaiting review"
state to the
.I "being reviewed"
state.
.\" ------------------------------------------------------------------------
.PS
boxwid = 1
down
S1: box "awaiting" "review"
arrow " review" ljust " begin" ljust
S2: box "being" "reviewed"
move to S2.w
T1: spline left 0.5 then up 1 then to S1.w ->
" review" ljust " begin" ljust " undo" ljust at T1.c-(0.5,0)
.PE
.\" ------------------------------------------------------------------------
.SS Notification
If the \fIreview_begin_notify_command\fP has been set in the project
attributes, this command will be run.  This is usually used to tell
other reviewers that you have started review, and they need not.
See \fIaepattr\fP(5) and \fIaepa\fP(1) for more information.
.PP
If used when the \fIdevelop_end_action\fP project attribute is set to
\fIgoto_being_reviewed\fP, then only the notification message is sent.
.SH OPTIONS
The following options are understood:
.so lib/en/man1/o_change.so
.so lib/en/man1/o_help.so
.so lib/en/man1/o_list.so
.so lib/en/man1/o_project.so
.so lib/en/man1/o_reason2.so
.so lib/en/man1/o_terse.so
.so lib/en/man1/o_verbose.so
.so lib/en/man1/o_wait.so
.so lib/en/man1/o__rules.so
.SH RECOMMENDED ALIAS
The recommended alias for this command is
.nf
.ta 8n 16n
csh%    alias aerb '\*(n) \-rb \e!* \-v'
sh$     aerb(){\*(n) \-rb "$@" \-v}
.fi
.SH ERRORS
It is an error if
the change is not in the
.I "awaiting review"
state.
.br
It is an error if
the current user is not a reviewer of the project.
.br
Its is an error if
the current user developed the change and the project is configured to
not permit developers to review their own changes (default).
.so lib/en/man1/z_exit.so
.br
.ne 1i
.SH SEE ALSO
.TP 8n
.IR aecd (1)
change directory
.TP 8n
.IR aede (1)
complete development of a change
.TP 8n
.IR aedeu (1)
recall a change for further development
.TP 8n
.IR aerpass (1)
pass review of a change
.TP 8n
.IR aeib (1)
begin integrating a change
.TP 8n
.IR aenrv (1)
add a reviewer to a project
.TP 8n
.IR aerfail (1)
fail review of a change
.TP 8n
.IR aerpu (1)
rescind a change review pass
.TP 8n
.IR aeuconf (5)
user configuration file format
.so lib/en/man1/z_cr.so
.\" vim: set ts=8 sw=4 et :


etc/CHANGES



Project "aegis"								  Page 1
Details							Sat Apr	 5 23:36:53 2003

NAME
	Project "aegis".

SUMMARY
	project change supervisor.

DESCRIPTION
	project change supervisor.

BRANCH CONTENTS
	Change	State		Description
	------- -------		-----------
	3	completed	project change supervisor, branch 3.
	4	being_developed project change supervisor, branch 4.

ARCHITECTURES
	This change must be developed on the unspecified architectures.

	arch.	 host	  aeb	      aet	  aet -bl     aet -reg
	-------- -------- ---------   ---------	  ---------   ---------
	unspecif	  required    required	  required    exempt
	ied

CAUSE
	This change was caused by internal_enhancement.

FILES
	Type	Action	 Edit	      File Name
	------	------	 ------	      ------
	build	remove	 5	      BUILDING
				      Deleted by change 3.
	source	remove	 2 (5)	      LICENSE
				      Deleted by change 3.
	build	remove	 8 (5)	      MANIFEST
				      Deleted by change 3.
	build	remove	 10 (5)	      Makefile.in
				      Deleted by change 3.
	build	remove	 6 (5)	      README
				      Deleted by change 3.
	source	create	 7 (5)	      aedist/arglex3.c
	source	create	 7 (5)	      aedist/arglex3.h
	source	remove	 4 (5)	      aedist/header.c
				      Deleted by change 3.
				      Moved to libaegis/rfc822header.c
	source	remove	 1 (5)	      aedist/header.h
				      Deleted by change 3.
				      Moved to libaegis/rfc822header.h
	source	remove	 5	      aedist/input/822wrap.c
				      Deleted by change 3.
				      Moved to libaegis/input/822wrap.c
	source	remove	 2 (5)	      aedist/input/822wrap.h
				      Deleted by change 3.
				      Moved to libaegis/input/822wrap.h
	source	create	 3 (5)	      aedist/input/cpio.c
	source	create	 1 (5)	      aedist/input/cpio.h
	source	create	 5	      aedist/input/cpio_child.c
	source	create	 1 (5)	      aedist/input/cpio_child.h



Project "aegis"								  Page 2
Details							Sat Apr	 5 23:36:53 2003

	Type	Action	 Edit	      File Name
	------	------	 ------	      ------
	source	create	 9 (5)	      aedist/list.c
	source	create	 1 (5)	      aedist/list.h
	source	create	 5	      aedist/main.c
	source	create	 7 (5)	      aedist/open.c
	source	create	 2 (5)	      aedist/open.h
	source	create	 4 (5)	      aedist/output/cpio.c
	source	create	 1 (5)	      aedist/output/cpio.h
	source	create	 5	      aedist/output/cpio_child.c
	source	create	 1 (5)	      aedist/output/cpio_child.h
	source	create	 4 (5)	      aedist/output/cpio_child2.c
	source	create	 1 (5)	      aedist/output/cpio_child2.h
	source	create	 29 (5)	      aedist/receive.c
	source	create	 1 (5)	      aedist/receive.h
	source	create	 21 (5)	      aedist/send.c
	source	create	 1 (5)	      aedist/send.h
	source	create	 2 (5)	      aefind/cmdline.h
	source	create	 8 (5)	      aefind/cmdline.y
	source	create	 4 (5)	      aefind/descend.c
	source	create	 2 (5)	      aefind/descend.h
	source	create	 3 (5)	      aefind/function.c
	source	create	 2 (5)	      aefind/function.h
	source	create	 2 (5)	      aefind/function/basename.c
	source	create	 2 (5)	      aefind/function/basename.h
	source	create	 3 (5)	      aefind/function/needs.c
	source	create	 2 (5)	      aefind/function/needs.h
	source	create	 2 (5)	      aefind/function/print.c
	source	create	 2 (5)	      aefind/function/print.h
	source	create	 4 (5)	      aefind/function/stat.c
	source	create	 2 (5)	      aefind/function/stat.h
	source	create	 5	      aefind/lex.c
	source	create	 2 (5)	      aefind/lex.h
	source	create	 2 (5)	      aefind/main.c
	source	create	 2 (5)	      aefind/shorthand/name.c
	source	create	 2 (5)	      aefind/shorthand/name.h
	source	create	 2 (5)	      aefind/shorthand/path.c
	source	create	 2 (5)	      aefind/shorthand/path.h
	source	create	 2 (5)	      aefind/shorthand/print.c
	source	create	 2 (5)	      aefind/shorthand/print.h
	source	create	 3 (5)	      aefind/shorthand/stat.c
	source	create	 2 (5)	      aefind/shorthand/stat.h
	source	create	 2 (5)	      aefind/tree.c
	source	create	 2 (5)	      aefind/tree.h
	source	create	 4 (5)	      aefind/tree/arithmetic.c
	source	create	 2 (5)	      aefind/tree/arithmetic.h
	source	create	 3 (5)	      aefind/tree/bitwise.c
	source	create	 2 (5)	      aefind/tree/bitwise.h
	source	create	 2 (5)	      aefind/tree/constant.c
	source	create	 2 (5)	      aefind/tree/constant.h
	source	create	 2 (5)	      aefind/tree/diadic.c
	source	create	 2 (5)	      aefind/tree/diadic.h
	source	create	 2 (5)	      aefind/tree/list.c
	source	create	 2 (5)	      aefind/tree/list.h
	source	create	 3 (5)	      aefind/tree/logical.c
	source	create	 2 (5)	      aefind/tree/logical.h
	source	create	 3 (5)	      aefind/tree/match.c



Project "aegis"								  Page 3
Details							Sat Apr	 5 23:36:53 2003

	Type	Action	 Edit	      File Name
	------	------	 ------	      ------
	source	create	 2 (5)	      aefind/tree/match.h
	source	create	 2 (5)	      aefind/tree/monadic.c
	source	create	 2 (5)	      aefind/tree/monadic.h
	source	create	 2 (5)	      aefind/tree/now.c
	source	create	 2 (5)	      aefind/tree/now.h
	source	create	 2 (5)	      aefind/tree/private.c
	source	create	 2 (5)	      aefind/tree/private.h
	source	create	 4 (5)	      aefind/tree/relative.c
	source	create	 2 (5)	      aefind/tree/relative.h
	source	create	 2 (5)	      aefind/tree/this.c
	source	create	 2 (5)	      aefind/tree/this.h
	source	create	 3 (5)	      aefp/main.c
	source	create	 17 (5)	      aegis/aeb.c
	source	create	 2 (5)	      aegis/aeb.h
	source	create	 13 (5)	      aegis/aeca.c
	source	create	 3 (5)	      aegis/aeca.h
	source	create	 10 (5)	      aegis/aecd.c
	source	create	 2 (5)	      aegis/aecd.h
	source	create	 11 (5)	      aegis/aechown.c
	source	create	 2 (5)	      aegis/aechown.h
	source	create	 12 (5)	      aegis/aeclean.c
	source	create	 1 (5)	      aegis/aeclean.h
	source	create	 19 (5)	      aegis/aeclone.c
	source	create	 1 (5)	      aegis/aeclone.h
	source	create	 19 (5)	      aegis/aecp.c
	source	create	 2 (5)	      aegis/aecp.h

aegis/aerb.cc
//
// aegis - project change supervisor
// Copyright (C) 2001-2008, 2011, 2012 Peter Miller
// Copyright (C) 2008 Walter Franzini
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3 of the License, or (at
// your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//

#include <common/ac/stdio.h>
#include <common/ac/stdlib.h>
#include <common/ac/sys/types.h>
#include <common/ac/sys/stat.h>

#include <common/mem.h>
#include <common/progname.h>
#include <common/quit.h>
#include <common/sizeof.h>
#include <common/trace.h>
#include <libaegis/ael/change/by_state.h>
#include <libaegis/arglex2.h>
#include <libaegis/arglex/change.h>
#include <libaegis/arglex/project.h>
#include <libaegis/change/file.h>
#include <libaegis/change/identifier.h>
#include <libaegis/change.h>
#include <libaegis/commit.h>
#include <libaegis/dir.h>
#include <libaegis/file.h>
#include <libaegis/help.h>
#include <libaegis/lock.h>
#include <libaegis/os.h>
#include <libaegis/pattr.fmtgen.h>
#include <libaegis/project/file.h>
#include <libaegis/project.h>
#include <libaegis/project/history.h>
#include <libaegis/rss.h>
#include <libaegis/sub.h>
#include <libaegis/undo.h>
#include <libaegis/user.h>

#include <aegis/aerb.h>


static void
review_begin_usage(void)
{
    const char      *progname;

    progname = progname_get();
    fprintf
    (
        stderr,
        "usage: %s -Review_Begin <change_number> [ <option>... ]\n",
        progname
    );
    fprintf
    (
        stderr,
        "       %s -Review_Begin -List [ <option>... ]\n",
        progname
    );
    fprintf(stderr, "       %s -Review_Begin -Help\n", progname);
    quit(1);
}


static void
review_begin_help(void)
{
    help("aerb", review_begin_usage);
}


static void
review_begin_list(void)
{
    trace(("review_begin_list()\n{\n"));
    arglex();
    change_identifier cid;
    cid.command_line_parse_rest(review_begin_usage);
    int mask = 1 << cstate_state_awaiting_review;

    //
    // It's messy.  Depending on the "develop end action" project
    // attribute, we have to cope with both "awaiting review" and
    // "being reviewed" states.
    //
    // This is because when the action is "goto_being_reviewed"
    // then the aerb command still runs the "review_begin_notify_
    // command", which is usually used to send email (or news)
    // to other reviewers and thus avoid duplicating effort.
    //
    if
    (
        project_develop_end_action_get(cid.get_pp())
    ==
        pattr_develop_end_action_goto_being_reviewed
    )
    {
        mask |= 1 << cstate_state_being_reviewed;
    }

    list_changes_in_state_mask(cid, mask);
    trace(("}\n"));
}


static void
review_begin_main(void)
{
    cstate_ty       *cstate_data;
    cstate_history_ty *history_data;
    long            j;

    trace(("review_begin_main()\n{\n"));
    arglex();
    change_identifier cid;
    string_ty *reason = 0;
    while (arglex_token != arglex_token_eoln)
    {
        switch (arglex_token)
        {
        default:
            generic_argument(review_begin_usage);
            continue;

        case arglex_token_change:
        case arglex_token_number:
        case arglex_token_project:
        case arglex_token_string:
            cid.command_line_parse(review_begin_usage);
            continue;

        case arglex_token_wait:
        case arglex_token_wait_not:
            user_ty::lock_wait_argument(review_begin_usage);
            break;

        case arglex_token_reason:
            if (reason)
                duplicate_option(review_begin_usage);
            switch (arglex())
            {
            default:
                option_needs_string(arglex_token_reason, review_begin_usage);
                // NOTREACHED

            case arglex_token_string:
            case arglex_token_number:
                reason = str_from_c(arglex_value.alv_string);
                break;
            }
            break;
        }
        arglex();
    }
    cid.command_line_check(review_begin_usage);

    project *pp = cid.get_pp();
    user_ty::pointer up = cid.get_up();
    change::pointer cp = cid.get_cp();

    //
    // lock the change for writing
    //
    change_cstate_lock_prepare(cp);
    lock_take();
    cstate_data = cp->cstate_get();

    //
    // it is not an error if
    //      - the change is in the 'awaiting_review' state, OR
    //      - the "develop end action" is "goto being reviewed"
    //        and the change is in the 'being_reviewed' state.
    //
    if
    (
        cstate_data->state != cstate_state_awaiting_review
    &&
        (
                project_develop_end_action_get(pp)
            !=
                pattr_develop_end_action_goto_being_reviewed
        ||
            cstate_data->state != cstate_state_being_reviewed
        )
    )
        change_fatal(cp, 0, i18n("bad rb state"));
    if (!project_reviewer_query(pp, up->name()))
        project_fatal(pp, 0, i18n("not a reviewer"));
    if
    (
        !project_developer_may_review_get(pp)
    &&
        nstring(cp->developer_name()) == up->name()
    )
        change_fatal(cp, 0, i18n("developer may not review"));

    //
    // change the state
    // remember who reviewed it
    // add to the change's history
    //
    // Don't do anything if the change is already in the being_reviewed state.
    //
    if (cstate_data->state != cstate_state_being_reviewed)
    {
        cstate_data->state = cstate_state_being_reviewed;
        history_data = change_history_new(cp, up);
        history_data->what = cstate_history_what_review_begin;
        history_data->why = reason;
    }

    //
    // It is an error if any of the change files have been tampered
    // with, or any of the difference files.
    //
    for (j = 0;; ++j)
    {
        fstate_src_ty   *src_data;
        string_ty       *path;
        string_ty       *path_d;
        int             same;
        int             file_required;

        src_data = change_file_nth(cp, j, view_path_first);
        if (!src_data)
            break;

        file_required = 1;
        bool diff_file_required = change_diff_required(cp);
        switch (src_data->usage)
        {
        case file_usage_build:
            file_required = 0;
            diff_file_required = false;
            break;

        case file_usage_source:
        case file_usage_config:
        case file_usage_test:
        case file_usage_manual_test:
            break;
        }
        switch (src_data->action)
        {
        case file_action_remove:
            file_required = 0;

            //
            // the removed half of a move is not differenced
            //
            if
            (
                src_data->move
            &&
                cp->file_find(nstring(src_data->move), view_path_first)
            )
                diff_file_required = false;
            break;

        case file_action_create:
        case file_action_modify:
        case file_action_insulate:
        case file_action_transparent:
            break;
        }

        path = cp->file_path(src_data->file_name);
        if (file_required)
        {
            up->become_begin();
            same = change_fingerprint_same(src_data->file_fp, path, 0);
            up->become_end();
            if (!same)
            {
                sub_context_ty  *scp;

                scp = sub_context_new();
                sub_var_set_string(scp, "File_Name", src_data->file_name);
                change_fatal(cp, scp, i18n("$filename altered"));
                // NOTREACHED
                sub_context_delete(scp);
            }
        }

        path_d = str_format("%s,D", path->str_text);
        trace_bool(diff_file_required);
        if (diff_file_required)
        {
            up->become_begin();
            same = change_fingerprint_same(src_data->diff_file_fp, path_d, 0);
            up->become_end();
            if (!same)
            {
                sub_context_ty  *scp;

                scp = sub_context_new();
                sub_var_set_format
                (
                    scp,
                    "File_Name",
                    "%s,D",
                    src_data->file_name->str_text
                );
                change_fatal(cp, scp, i18n("$filename altered"));
                // NOTREACHED
                sub_context_delete(scp);
            }
        }
        str_free(path);
        str_free(path_d);
    }

    //
    // write out the data and release the locks
    //
    cp->cstate_write();
    commit();
    lock_release();

    //
    // run the notify command
    //
    cp->run_review_begin_notify_command();

    //
    // Update the RSS feed file if necessary.
    //
    rss_add_item_by_change(pp, cp);

    //
    // verbose success message
    //
    change_verbose(cp, 0, i18n("review begin complete"));
    change_free(cp);
    project_free(pp);
    trace(("}\n"));
}


void
review_begin(void)
{
    static arglex_dispatch_ty dispatch[] =
    {
        { arglex_token_help, review_begin_help, 0 },
        { arglex_token_list, review_begin_list, 0 },
    };

    trace(("review_begin()\n{\n"));
    arglex_dispatch(dispatch, SIZEOF(dispatch), review_begin_main);
    trace(("}\n"));
}


// vim: set ts=8 sw=4 et :
